### DISCUSSION TITLE
'0x00' in UART buffer before writing anything.

### DISCUSSION THREAD

**Discussion by lhchang** [micropython/micropython#12217] [RP2040 / Pico]
I've been trying to debug not receiving anything over UART when I write something and among other things encountering an odd situation where upon program start, UART.any() always returns 1 and the value read is always `b'0x00'` prior to me sending anything at all. I was wondering if I am misconfiguration anything that's causing this to happen? For reference I a have just been trying this in the REPL shell (https://github.com/orgs/micropython/discussions/11848#discussioncomment-6325045) for testing serial communications. Thank you

**Comment by robert-hh** [comment]
You did not mention the MicroPython version. And you should show the code sequence you use.
Generally, it may happen that one read some characters after initializing the UART. Any low level at the RX pin will be considered as start bit, and then a data byte is read by the UART. It may happen the the RX pin is low when UART is set up, at least for a short while.

**Reply by peterhinch** [reply]
@lhchang If you aren't using the receiver the Rxd should be pulled up if you want to prevent this. As @robert-hh has pointed out, the default state on a UART line is high.

**Reply by lhchang** [reply]
I apologize - I am running MicroPython v1.20.0. Code sequence I copy-pasted using @shariltumin 's example

```
MicroPython v1.20.0-242-g6a9db521e on 2023-06-30; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> from machine import UART, Pin
>>> uart = UART(0, baudrate=9600, tx=Pin(0), rx=Pin(1))
>>> uart.write('hello world')
11
>>> uart.write('hello world')
11
>>> w = uart.read()
>>> print(w)
b'hello worldhello world'
>>>
```

**Comment by shariltumin** [comment]
I assume you did the test on a Pi Pico.

```
>>> 
MPY: soft reboot
MicroPython v1.20.0 on 2023-04-26; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> import sys
>>> sys.platform
'rp2'
>>> from machine import UART, Pin
>>> uart = UART(1, baudrate=9600, tx=Pin(4), rx=Pin(5))
>>> uart.any()
0
>>> uart.write('hello world')
11
>>> uart.any()
11
>>> w = uart.read()
>>> print(w)
b'hello world'
>>> uart.any()
0
>>> uart.deinit()
>>> uart.any()
256
>>> w = uart.read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError: memory allocation failed, allocating 183808 bytes
>>> uart.init()
>>> uart.any()
0
>>> w = uart.read()
>>> w
>>>
```

------------------------------------------------------

rp2 uart pins [see](https://docs.micropython.org/en/latest/_images/pico_pinout.png)

| uart | tx (GPIO) | rx (GPIO)|
|:----:|:---------:|:--------:|
| 0    | 0         | 1        |
| 0    | 12        | 13       |
| 0    | 16        | 17       |
| 1    | 4         | 5        |
| 1    | 8         | 9        |

The pin numbers are the GPIO numbers, not the PIN labels, i.e. for UART1(4,5) connect PIN labelled 6 and 7 together (I used a paper clip!).

Now, what happens if I disconnect the two GPIOs?

```
>>> uart.any()
1
>>> w = uart.read()
>>> w
b'\x00'
>>>
```

You may want to check your connection or try different PIN connections.

**Reply by lhchang** [reply]
Yes, this is the test I performed. I think you are right that the connection is broken: along with getting the same result you had in your second code snippet, I am still unable to read anything from the buffer after reading that initial b'\x00 and then writing something else. Thanks for the hint, will check it out here soon and hopefully fix the issue.

**Comment by robert-hh** [comment]
Sorry if I ask for an obvious topic, but did you connect the GND of both devices as well as TX and TX.

**Comment by shariltumin** [comment]
Robert, in my test the two gpio pins are on the same board. I just short-circuited them. This is just to test if the machine.UART() works or not. So I suggest trying different legal GPIO combinations.

**Comment by kjm1102** [comment]
After power up I often find junk in the Rx buffer, even when the 2 uarts are on the same chip. I just do a uart.read() before any write to make sure the buffer is clear.

**Reply by shariltumin** [reply]
@kjm1102, You may want to try ```uart.deinit()``` and  ```uart.init()``` after power up.

**Comment by peterhinch** [comment]
> Now, what happens if I disconnect the two GPIOs?
```
>>> uart.any()
1
>>> w = uart.read()
>>> w
b'\x00'
>>>
```
As I tried to explain before, in this test the Rxd pin is floating, which is not a valid condition. It should be pulled up to 3.3V, for example with a resistor. UART lines should be high in the idle state otherwise garbage will be read.

**Comment by shariltumin** [comment]
Exactly! The point I want to make is that a wrong connection has been made. For example, PIN#4 and PIN#5 have been connected, which are not GPIO(4) and GPIO(5), which are PIN#6 and PIN#7. According to [pico_pinout](https://docs.micropython.org/en/latest/_images/pico_pinout.png)

```
>>> uart = UART(1, baudrate=9600, tx=Pin(4), rx=Pin(5))
>>> # no short
>>> Pin(4).value()
1
>>> Pin(5).value()
0
>>> # short Pin(4) and Pin(5) on the same board!
>>> Pin(4).value()
1
>>> Pin(5).value()
1
>>> as expected!
>>> # say by mistake we short PIN#4(GPIO2) and PIN#5(GPIO3)
>>> Pin(2).value()
0
>>> Pin(3).value()
0
>>> # uart.write() does not write to Pin(2) and uart.read() does not read Pin(3).
>>> # both pins are floating at 0.
>>> # reconnect Pin(4) and Pin(5)
>>> w = uart.read()
>>> print(w)
b'\x00\x00\xfe\x0fu\x04\x80a\x08\x00\xe0>\xff\xff\x00\x00\x00\x1f\x00\xf5\x00'
>>> w = uart.read()
>>> print(w)
None
>>> # I get junk for the 1st read
>>> # and for the 2.nd read, I get None
>>> # as expected
```

Both RP-Pico and RP-Pico-W boards have GPIO numbers on the bottom. It is not confusing. I do not know what kind of RP2040 board @Ihchang has been trying the test on.
